### PR TITLE
WIP: DH-12293 Utilities to assist in retrieving data from tables

### PR DESCRIPTION
**Pull request by rbasralian**
Adds utilities to make it easier to pull data from a table.

There is a `KeyedRecordAdapter` that leverages the ToMapListener to allow pulling data by key (e.g. USym):

```java
final KeyedRecordAdapter<String, MyTradeHolder> keyedRecordAdapter = KeyedRecordAdapter.makeKeyedRecordAdapter(
                tradesTable,
                myTradeHolderRecordAdapterDescriptor,
                "USym",
                String.class
        );
final MyTradeHolder lastAaplTrade = keyedRecordAdapter.getRecord("AAPL");
final Map<String, MyTradeHolder> lastTradesBySymbol = keyedRecordAdapter.getRecords("AAPL", "CAT", "SPY");
```

And a `TableToRecordListener` that lets you listen to updates as objects instead of just row keys:

```java
final Consumer<MyTradeHolder> tradeConsumer = trade -> System.out.println(trade.toString());
TableToRecordListener.create(tradesTable, myTradeHolderRecordAdapterDescriptor, tradeConsumer);
```

The mapping of columns to fields in whatever object is handled by a RecordAdapterDescriptor:

```java
RecordAdapterDescriptor<MyTradeHolder> myTradeHolderRecordAdapterDescriptor = RecordAdapterDescriptorBuilder
            .create(MyTradeHolder::new)
            .addColumnAdapter("Sym", RecordUpdater.getStringUpdater(MyTradeHolder::setSym))
            .addColumnAdapter("Price", RecordUpdater.getDoubleUpdater(MyTradeHolder::setPrice))
            .addColumnAdapter("Size", RecordUpdater.getIntUpdater(MyTradeHolder::setSize))
            .addColumnAdapter("Timestamp", RecordUpdater.getReferenceTypeUpdater(DBDateTime.class, MyTradeHolder::setTimestamp))
            .build();
```

The RecordAdapterDescriptor can be used to create a `SingleRowRecordAdapter` or a `MultiRowRecordAdapter`. A single-row adapter creates an object (`MyTradeHolder` or `JsonNode` or whatever) and reads data from the columns directly into that object. A multi-row adapter has more steps but is intended to be efficient and chunk-oriented. It will:

1. Create arrays to hold data for all of the columns.
2. Read the data for all columns into those arrays (by chunk).
3. Create an array of records (e.g. `MyTradeHolder[]`, `JsonNode[]`, whatever) and fill it with new empty records.
4. Populate all of the records with the data from the arrays, one column at a time.

A record adapter instance can be generated based on the descriptor. For example, `com.illumon.iris.db.util.dataadapter.rec.json.JsonRecordAdapterUtil#createJsonRecordAdapterDescriptor(t, "Col1", "Col2", "Col3").createMultiRowRecordAdapter(t)` would figure out the types of "Col1"/"Col2"/"Col3" and generate a class that populates ObjectNodes with those fields.

There is also a draft Python version that lets you create data structures by passing data to either a function or straight to some type's constructor:

```python
class StockTrade:
    sym: str
    price: float
    size: int
    exch: str

    def __init__(self, sym: str, price: float, size: int, exch: str):
        self.sym = sym
        self.price = price
        self.size = size
        self.exch = exch

keyed_record_adapter: KeyedRecordAdapter[str, StockTrade] = KeyedRecordAdapter(
    source,
    StockTrade,
    'Sym',
    ["Price", "Size", "Exch"]
)

records = keyed_record_adapter.get_records(["AAPL", None])

aapl_trade = records['AAPL']
self.assertEquals('AAPL', aapl_trade.sym)
self.assertEquals(1.1, aapl_trade.price)
self.assertEquals(10_000_000_000, aapl_trade.size)
self.assertEquals('ARCA', aapl_trade.exch)
```

Random notes:
- This is a generic version of a utility we previously used for latency monitoring (that's where the "create arrays to hold data, consistently read chunks from columns and copy the chunks to the arrays, then copy the data from the arrays back into row-oriented data structures" part comes from)
- The `MultiRowRecordAdapter` is responsible for the real work of pulling data out of tables; both `KeyedRecordAdapter` and `TableToRecordListener` use that. (`KeyedRecordAdapter` also uses `SingleRowRecordAdapter` when fetching single keys.)
- The Python version (`keyed_record_adapter.py`/`PyKeyedRecordAdapter.java`) only supports strings as keys and has no listening component.
- (The Java unit tests pass. The Python unit tests fail on primitive keys.)

### CODE REVIEW

**Comment by margaretkennedy** [comment]
https://github.com/deephaven/deephaven.io/issues/1715

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+        # Index of current row (i.e. index into arrays of data retrieved from each DB column)
+        row_idx = 0
+        for row_key in record_data_row_keys:
+            # Find the data key corresponding to the DB row key from
+            # which this record was retrieved. Map the data key to the record.
+            data_key_idx: int = row_key_to_data_key_positional_index.get(row_key)
+            data_key_for_row = data_keys[data_key_idx]
+
+            # print(f'Building record for row_idx {row_idx}, row_key {row_key}, data_key {data_key_idx}/{data_key_for_row}')
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:110]: remove?

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+            row_data = list()
+            if self.is_single_key_col:
+                row_data.append(data_key_for_row)
+            else:
+                row_data.extend(data_key_for_row)
+
+            for col_arr in record_data_arrs:
+                # print(f'Adding data: {col_arr[row_idx]}')
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:119]: remove?

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                row_data.append(col_arr[row_idx])
+
+            print(f'row_data for row {row_idx} is: {row_data}')
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:122]: remove?

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                    results_list.append(None)
+
+            return results_list
+
+    def _convert_keys_seq_for_java(self, data_keys: typing.Sequence, key_dtype: DType) -> typing.Sequence:
+        if type(data_keys) is numpy.ndarray:
+            # Some types (e.g. bool/DateTime) have special handling for numpy:
+            return dhnp._np_to_dh_seq(key_dtype, data_keys)
+        else:
+            return data_keys
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:153]: Can't `dhnp._np_to_dh_seq(` just be used in place of this whole function, since it returns the original value if it doesn't need to convert?

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+def make_record_adapter(table: Table,
+                        record_adapter_function: typing.Callable[..., T],
+                        key_cols: [str, typing.List[str]],
+                        data_cols: [str, typing.List[str]]) -> KeyedRecordAdapter[K, T]:
+    return KeyedRecordAdapter(
+        table,
+        record_adapter_function,
+        key_cols,
+        data_cols
+    )
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:184]: It is more pythonic to condense these two functions into just one `make_record_adapter` that inspects the input values to construct this in the desired way.

File: py/server/deephaven/experimental/keyed_record_adapter.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+def make_record_adapter_with_constructor(table: Table,
+                                         record_adapter_type: typing.Type[T],
+                                         key_cols: [str, typing.List[str]],
+                                         data_cols: [str, typing.List[str]]) -> KeyedRecordAdapter[K, T]:
+    def constructor_adapter_function(*row_data):
+        # Create the record from the data key and row data
+        # The constructor MUST take the keys + the data as arguments, in order.
+        record = record_adapter_type.__new__(record_adapter_type)
+        record_adapter_type.__init__(record, *row_data)
+        return record
```
Comment (chipkent) [inline py/server/deephaven/experimental/keyed_record_adapter.py:165]: Is there a reason that you aren't just doing
`return record_adapter_type(record, *row_data)`?  ... which could be simplified to a lambda.

image

File: py/server/tests/test_keyed_record_adapter.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+class KraTestCase(BaseTestCase):
+
+    def testCustomKeyedRecordAdapterWithOneStrKeyCol(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", None),
+        # ))
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:24]: Kill?

File: py/server/tests/test_keyed_record_adapter.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            size: int
+            exch: str
+
+            def __init__(self, sym: str, price: float, size: int, exch: str):
+                self.sym = sym
+                self.price = price
+                self.size = size
+                self.exch = exch
+
+        keyed_record_adapter: KeyedRecordAdapter[str, StockTrade] = KeyedRecordAdapter(
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:46]: type hints are not normally included in a place like this, since they can be inferred.

File: py/server/tests/test_keyed_record_adapter.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        self.assertEquals('ARCA', aapl_trade.exch)
+
+    def testCustomKeyedRecordAdapterWithOneIntKeyCol(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", None),
+        # ))
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:71]: Kill?

File: py/server/tests/test_keyed_record_adapter.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+        self.assertEquals('ARCA', aapl_trade.exch)
+
+    def testCustomKeyedRecordAdapterWithStrIntKeyCols(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, 1.2, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, 1_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", 'NASDAQ', None),
+        # ))
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:118]: Kill?

File: py/server/tests/test_keyed_record_adapter.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+            size: int
+
+            def __init__(self, sym: str, exch: int, price: float, size: int, ):
+                self.sym = sym
+                self.exch = exch
+                self.price = price
+                self.size = size
+
+        keyed_record_adapter: KeyedRecordAdapter[
+            List[str], StockTradeIntExchange] = make_record_adapter_with_constructor(
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:141]: type hints probably not needed

File: py/server/tests/test_keyed_record_adapter.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        self.assertEquals(None, null_trade.exch)
+
+    def testCustomKeyedRecordAdapterWithStrStrKeyCols(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, 1.2, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, 1_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", 'NASDAQ', None),
+        # ))
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:182]: Kill?

File: py/server/tests/test_keyed_record_adapter.py [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+            size: int
+
+            def __init__(self, sym: str, exch: str, price: float, size: int, ):
+                self.sym = sym
+                self.exch = exch
+                self.price = price
+                self.size = size
+
+        keyed_record_adapter: KeyedRecordAdapter[
+            List[str], StockTradeAlternateConstructor] = make_record_adapter_with_constructor(
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:205]: unneeded type hint?

File: py/server/tests/test_keyed_record_adapter.py [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+        self.assertEquals(None, records[2])
+
+    def testCustomKeyedRecordAdapterWithCompositeKeyColAndAdapter(self):
+        # source = Table(_JTstTools.testRefreshingTable(
+        #     _JTstTools.i(2, 4).copy().toTracking(),
+        #     _JTableTools.intCol("Sym", 'AAPL', 'AAPL', None),
+        #     _JTableTools.doubleCol("Price", 1.1, 1.2, NULL_DOUBLE),
+        #     _JTableTools.longCol("Size", 10_000_000_000, 1_000, NULL_LONG),
+        #     _JTableTools.col("Exch", "ARCA", 'NASDAQ', None),
+        # ))
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:242]: Kill?

File: py/server/tests/test_keyed_record_adapter.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+                self.price = price
+                self.size = size
+                self.exch = exch
+
+        # Function that just calls StockTrades() with args in correct order (Exch last; when calling
+        # this method it is second arg because keys come first):
+        def create_stock_trade(sym: str, exch: str, price: float, size: int):
+            return StockTrade(sym, price, size, exch)
+
+        keyed_record_adapter: KeyedRecordAdapter[List[str], StockTrade] = make_record_adapter(
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:269]: unneeded type hint?

File: py/server/tests/test_keyed_record_adapter.py [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+        self.assertEquals('AAPL', aapl_trade.sym)
+        self.assertEquals(1.2, aapl_trade.price)
+        self.assertEquals(1_000, aapl_trade.size)
+        self.assertEquals('NASDAQ', aapl_trade.exch)
+
+        self.assertEquals(None, records[2])
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (chipkent) [inline py/server/tests/test_keyed_record_adapter.py:301]: In PartitionedTables, @rcaudy has resolved the 1- vs many-key values problem.  This whole PR should strive for the same for a user experience.

**Review by chipkent — CHANGES_REQUESTED** [review]
Overall, I found this PR very tough to review in a thorough way.
1) There is a ton of boilerplate that obscures the signal I'm trying to see.
2) The overall high-level design for the Java code is not apparent, which makes it hard to follow.
3) I have a feeling that the java could be simplified in some significant way, but I'm at a loss for saying what this would look like, since I'm still not fully grasping the high-level view.

File: QueryUtil/src/main/java/io/deephaven/queryutil/dataadapter/datafetch/bulk/gen/TableDataRetrieverGenerator.java [hunk 17]
```diff
@@ -0,0 +1,177 @@
+package io.deephaven.queryutil.dataadapter.datafetch.bulk.gen;
```
[the remark this answers is not in the document]
Comment (rbasralian) [reply QueryUtil/src/main/java/io/deephaven/queryutil/dataadapter/datafetch/bulk/gen/TableDataRetrieverGenerator.java:1]: it's not generated; it handles generation

File: QueryUtil/src/main/java/io/deephaven/queryutil/dataadapter/ChunkRetrievalUtil.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+import io.deephaven.chunk.*;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.table.ChunkSource;
+import io.deephaven.engine.table.ColumnSource;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+
+public class ChunkRetrievalUtil {
```
[the remark this answers is not in the document]
Comment (rbasralian) [reply QueryUtil/src/main/java/io/deephaven/queryutil/dataadapter/ChunkRetrievalUtil.java]: turned down this & a few others

**Comment by chipkent** [comment]
I have not looked at this PR recently, so I don't remember the contents.  DHC has added functionality to help users extract data.  I'm noting it here in case it makes this PR no longer needed.

https://github.com/deephaven/deephaven-core/pull/5595
https://deephaven.io/core/docs/how-to-guides/iterate-table-data/
